dartagnan/src/main/java/com/dat3m/dartagnan/analysis/Refinement.java

```diff
@@ -0,0 +1,353 @@
+        }
+
+
+        // This method computes a refinement clause from a set of violations.
+        // Furthermore, it computes symmetric violations if symmetry learning is enabled.
+        public BooleanFormula refine(DNF<CoreLiteral> coreViolations) {
+            BooleanFormulaManager bmgr = context.getFormulaManager().getBooleanFormulaManager();
+            BooleanFormula refinement = bmgr.makeTrue();
+            // For each symmetry permutation, we will create refinement clauses
+            for (Function<Event, Event> perm : symmPermutations) {
+                for (Conjunction<CoreLiteral> violation : coreViolations.getCubes()) {
+                    BooleanFormula permutedClause = bmgr.makeFalse();
+                    for (CoreLiteral literal : violation.getLiterals()) {
+                        permutedClause = bmgr.or(permutedClause, bmgr.not(permuteAndConvert(literal, perm)));
+                    }
+                    refinement = bmgr.and(refinement, permutedClause);
+                }
+            }
+            return refinement;
+        }
+
+
+
+        // Computes a list of permutations allowed by the program.
+        // Depending on the <learningOption>, the set of computed permutations differs.
+        // In particular, for the option NONE, only the identity permutation will be returned.
+        private List<Function<Event, Event>> computeSymmetryPermutations(SymmetryLearning learningOption) {
+            ThreadSymmetry symm = task.getThreadSymmetry();
+            Set<? extends EquivalenceClass<Thread>> symmClasses = symm.getNonTrivialClasses();
+            List<Function<Event, Event>> perms = new ArrayList<>();
+            perms.add(Function.identity());
+
+            for (EquivalenceClass<Thread> c : symmClasses) {
+                List<Thread> threads = new ArrayList<>(c);
+                threads.sort(Comparator.comparingInt(Thread::getId));
+
+                switch (learningOption) {
+                    case NONE:
+                        break;
+                    case LINEAR:
+                        for (int i = 0; i < threads.size(); i++) {
+                            int j = (i + 1) % threads.size();
+                            perms.add(symm.createTransposition(threads.get(i), threads.get(j)));
+                        }
+                        break;
+                    case QUADRATIC:
+                        for (int i = 0; i < threads.size(); i++) {
+                            for (int j = i + 1; j < threads.size(); j++) {
+                                perms.add(symm.createTransposition(threads.get(i), threads.get(j)));
+                            }
+                        }
+                        break;
+                    case FULL:
+                        List<Function<Event, Event>> allPerms = symm.createAllPermutations(c);
+                        allPerms.remove(Function.identity()); // We avoid adding multiple identities
+                        perms.addAll(allPerms);
```

Comment:
Because that only works if we have a single symmetry class. If you have multiple symmetry classes, then you would overwrite the permutations if you do it your way. For each symmetry class we need to add all permutations but we need the identity only once.